File: algorithm1/recursion_and_backtracking/ts/MergeTwoSortedLists.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
  const head: ListNode = new ListNode();
  let current: ListNode = head;

  while (list1 && list2) {
    // list1 과 list2가 있는 동안
    if (list1.val < list2.val) {
      // list2의 val이 더 크면 
      current.next = list1;
      // 다음 head는 list1이 된다. (그 다음이 list2가 될 것이기 때문)
      list1 = list1.next;
      // list1은 list1.next가 된다.
    } else {
      current.next = list2;
      list2 = list2.next;
    }
    current = current.next;
  }
  current.next = list1 || list2;
  return head.next;
};

File: top_interview_questions_easy/merge_two_sorted_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
      // null pointer literal
      // ** null은 값이기 때문에 NULL 보다는 nullptr을 사용하는게 더 정확하기 때문에 권장 
      if(list1 == nullptr) return list2;
      if(list2 == nullptr) return list1;

      // 재귀를 통한 구현
      if(list1->val <= list2->val) {
        // 더 큰 값을 다음 Node로 넘긴다.
        ListNode* nextNode = mergeTwoLists(list1->next, list2);
        list1->next = nextNode;
        return list1;
      }else {
        ListNode* nextNode = mergeTwoLists(list1, list2->next);
        list2->next = nextNode;
        return list2;
      }
    }
};